src/middleware/authorization.ts

```diff
@@ -0,0 +1,84 @@
+    try {
+      // We get token from header
+      const authHeader = req.headers["authorization"];
+      if (!authHeader)
+       { return res.status(404).send({ message: "Token not found in Header" });}
+  
+      const token = authHeader.split(" ")[1];
+      if (!token)
+        {return res.status(401).send({ message: "Token not found in Header" });}
+      // We verify the token
+      const result = process.env.ACCESS_TOKEN_SECRET && jwt.verify(token, process.env.ACCESS_TOKEN_SECRET ) as jwt.JwtPayload;
+      if (!result) {return res.status(403).send({ message: "Token not verified" });}
+      // We get the user id from token and fetch the user
+      
+      const user = await userModel.findById(result?._id);
+      // We set the user obj in the req so the next call can use our user object.
+      if(user)
+        {req.user = user
+          // If all good, we move to the next function
+          return next();}
+          else
+      {return res.status(404).send({ message: "User having token not in DB" });}
+
+      
+    } catch (err: any) {
+      return res.status(400).send({ type: err.name, message: err.message });
+    }
+  };
+
+  export const authorizeAdmin = async (req: RequestWithUser, res: Response, next: NextFunction) => {
+    // Check if the role is Admin or Not
+    // To authenticate Admin access only
+    if (req.user.role === "Admin") {
+      next();
+    }
+    else 
+    {return res.status(403).send({
+      message: "Not Authorized, Only Admin can access this route",}
+    );
+  }
+};
+
+export const authorizeProjectManager = async (req: RequestWithUser, res: Response, next: NextFunction) => {
+  // Check if the role is Admin / Project Manager or Not
+  if (req.user.role === "Admin" || req.user.role === "Project Manager") {
+    next();
+  }
+  else 
+  {return res.status(403).send({
+    message: "Not Authorized, Only Admin / Project Manager can access this route",}
+  );
+}
+}
+
+export const authorizeTeamLead = async (req: RequestWithUser, res: Response, next: NextFunction) => {
+  // Check if the role is Admin / Project Manager / Team Lead or Not
+  if (req.user.role === "Admin" || req.user.role === "Project Manager" || req.user.role === "Team Lead") {
+    next();
+  }
+  else 
+  {return res.status(403).send({
+    message: "Not Authorized, Only Admin / Project Manager / Team Lead can access this route",}
+  );
+}
+}
+
+// comment: Do we need this?
+export const authorizeTeamMember = async (req: RequestWithUser, res: Response, next: NextFunction) => { // comment: Not sure if we need this
+  // Check if the role is Admin / Project Manager / Team Lead or Not
+  if (req.user.role === "Admin" || req.user.role === "Project Manager" || req.user.role === "Team Lead" || req.user.role === "Team Member") {
+    next();
+  }
+  else 
+  {return res.status(403).send({
+    message: "Not Authorized, Only Admin / Project Manager / Team Lead / Team Member can access this route",}
+  );
+}
+}
```

Comment:
I think after all the middlewares we have added above, this won't be needed. We will be left with user actions then. Which mean anything an authorized user can do.